### DISCUSSION TITLE
Intersection with 3D object

### DISCUSSION THREAD

**Discussion by alex-klimov** [NVlabs/instant-ngp#746] [Q&A] [answered]
For our task we need efficient way to determine intersection with 3D object inside a NERF scene. Example of inefficient approach - export mesh and use existing algorithms based on mesh.

I would greatly appreciate some advice from the experts on the approach to take to check if particular 3D object intersects with the objects inside NERF scene.

**Comment by Tom94** [comment, marked as answer]
Hi, you could generate a number of positions within your 3D object (or on the surface of it, if surface collisions are sufficient) and query the NeRF model for its density at those positions. If the density is larger than some threshold (you could use the same threshold we use in the occupancy grid), you know the objects intersect.

I suspect that a couple thousand--million points suffice for this check to be accurate, which can be evaluated in about a millisecond per million points.

**Reply by alex-klimov** [reply]
Sounds good!

Will I need to write a new method taking `get_density_on_grid` as a template:
https://github.com/NVlabs/instant-ngp/blob/master/src/testbed_nerf.cu#L3426

Or is there anything ready for this task?

**Reply by Tom94** [reply]
Adapting `get_density_on_grid` sounds like a solid plan, I think.

**Reply by alex-klimov** [reply]
Yes, working on it. Thank you!

**Reply by 3a1b2c3** [reply]
Feel like adding a pr?

**Reply by alex-klimov** [reply]
In the middle of multiple tasks. The code is not in good condition at the moment. Later on - all the changes we have done to NGP - we would love to get merged into the master to avoid conflicts later on.
